packages/flutter/lib/src/foundation/consolidate_response.dart

```diff
@@ -6,19 +6,48 @@ import 'dart:async';
 import 'dart:io';
 import 'dart:typed_data';
 
-/// Efficiently converts the response body of an [HttpClientResponse] into a [Uint8List].
+/// Signature for getting notified when chunks of bytes are received while
+/// consolidating the bytes of an [HttpClientResponse] into a [Uint8List].
+///
+/// The `cumulative` parameter will contain the total number of bytes received
+/// thus far.
+///
+/// The `total` parameter will contain the _expected_ total number of bytes to
+/// be received (extracted from the value of the `Content-Length` HTTP response
+/// header), or -1 if the size of the response body is not known in advance.
+/// Even if the parameter reports a non-negative value, it is not 100%
+/// trustworthy (e.g. when GZIP is involved or other cases where an
+/// intermediate transformer has been applied to the stream).
+///
+/// This is used in [consolidateHttpClientResponseBytes].
+typedef BytesReceivedCallback = void Function(int cumulative, int total);
+
+/// Efficiently converts the response body of an [HttpClientResponse] into a
+/// [Uint8List].
 ///
 /// The future returned will forward all errors emitted by [response].
-Future<Uint8List> consolidateHttpClientResponseBytes(HttpClientResponse response) {
+///
+/// The [onBytesReceived] callback, if specified, will be invoked for every
+/// chunk of bytes that are received while consolidating the response bytes.
+/// For more information on how to interpret the parameters to the callback,
+/// see the documentation on [BytesReceivedCallback].
+Future<Uint8List> consolidateHttpClientResponseBytes(
+  HttpClientResponse response, {
+  BytesReceivedCallback onBytesReceived,
+}) {
   // response.contentLength is not trustworthy when GZIP is involved
   // or other cases where an intermediate transformer has been applied
-  // to the stream.
+  // to the stream, so we manually count the bytes as they cross the wire.
   final Completer<Uint8List> completer = Completer<Uint8List>.sync();
   final List<List<int>> chunks = <List<int>>[];
+  final int expectedContentLength = response.contentLength;
```

Comment:
This will almost always be less than the actual content length - your progress bar will end up loading faster because the contentLength will usually be the gzip compressed length.  You should be able to notice this if you test a network download of this with something from Wikimedia for example.

Reply:
Fixed.